### ISSUE TITLE
Not checking ports in use correctly

### ISSUE THREAD

**Issue by vlarcta** [man-group/dtale#561]
Hello, I have found that dtale sometimes fails to locate ports in use (running via jupyterhub proxy). Seems like running connect_ex is not safe enough. Referring to https://github.com/man-group/dtale/blob/a76234b137da2b88e0d4770076e305a5de9d1c37/dtale/app.py#L522-L524 

It can fail on some ports, reason for which I do not know yet (will update if find what is so special about these ports) but I have created a test:

```
def is_port_in_use(port):
    import socket 
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        result = sock.bind(('', port))
        return False
    except:
        return True
    finally:
        sock.close()

for port in range(40000,41000):
    if not dtale_app.is_port_in_use(port):
        if is_port_in_use(port):
            print(port)
```

I am getting at least a few ports in this range which dtale cannot detect are not available. Running dtale on these ports results in 'Address already in use' error.

**Comment by aschonfeld** [comment]
Hmm, I am not able to get the same exception. I believe this is the [port](https://stackoverflow.com/questions/2470971/fast-way-to-test-if-a-port-is-in-use-using-python) I used to build that piece of code.

**Comment by vlarcta** [comment]
Thanks for the reference. Is that also the case if you increase the port range to 49000 ? In my case what happens is dtale_app.is_port_in_use only finds unavailable ports if they are used by some other service but does not detect ports that are unavailable for any other reason (suspecting a firewall). Haven't found the true cause of what makes ports unavailable but netstat does not show anything running on these ports. Following this logic the port is *not* in use but is not available for any other reason.

**Comment by aschonfeld** [comment]
So here's information on [how ports are chosen](https://github.com/man-group/dtale#startup-behavior).  You can change the logic about which ports D-Tale searches by specifying some environment variables.  By default it looks for ports between 40000 - 49000.

That being said, there is nothing stopping you from forcing D-Tale to use a specific port.  You can accomplish that by doing: `dtale.show(df, port=8081, force=True)`

Lastly, you mentioned jupyterhub.  If you run D-Tale with Jupyter Server Proxy then it should bypass the whole port search logic and attach D-Tale to the same process as your jupyter notebook.  Here's information how that works: [D-Tale w/ Jupyter Server Proxy](https://github.com/man-group/dtale#jupyterhub-w-jupyter-server-proxy)

Let me know if any of this helps.

**Comment by vlarcta** [comment]
Hi, Thanks for the reply. I am using it in a bit more complicated fashion than running in a specific notebook for a specific port so the only thing that would work for me would be to override the port searching function to detect available ports correctly. I am trying to highlight that this logic is not sufficient to reliably select a port for the app and make sure the app will successfully run. 

As for jupyter server proxy part I do not see how it bypasses the port search logic and I can confirm that it makes a call to a function dtale_app.is_port_in_use when calling dtale.show for the first time in the notebook. The initial problem I had was that my port 40002 was not available and dtale failed to realise this and still tried to attach the app to this port. When I tried to open 3 notebooks that run dtale (respectively got assigned 40000,40001 and 40002) the third one always failed. 

Will have a look around web, maybe will come up with a pull request to improve this part of logic. Ofc the piece of code I had does the trick of finding a free port but I am pretty sure there is a neater way to achieve this.

**Comment by aschonfeld** [comment]
You are 100% correct about that code being fired when jupyter proxy server is being used. My mistake 🤦‍♂️Sorry

**Comment by aschonfeld** [comment]
@vlarcta this update has been released in v1.57.0
